### ISSUE TITLE
Decide on syntax for compiler virtual dependencies

### ISSUE THREAD

**Issue by rgommers** [rgommers/peps#3]
We figured out virtual dependencies with a fixed name, like `openmp` or `blas`. But for compilers we didn't figure it out yet. Options:
- "virtual:compiler{'rust'}" ~(this is in the draft right now)~ _changed to `virtual:compiler/c` now_
- "virtual:compiler-rust"  (use separate names)
- ... ?

@pradyunsg WDYT?

**Comment by rgommers** [comment]
Comment by @jaimegp:

> `virtual` packages use PURL too right? So what about `virtual:compiler/c`. Some tools might want Clang specifically, with a certain version (even if we are not providing this info yet in the absence of PURL version ranges).

No, as it stands `virtual:` isn't using PURL; PURLs exclusively start with `pkg:`. If we're separating `compiler` and `c`, then `compiler` looks like a "type" in:
```
scheme:type/namespace/name
```

Maybe there's something to that, we have these components:
- `scheme`: `pkg` or `virtual`
- `type`: `pypi`, `cargo`, `github`, `generic`, etc. for PURL. The `compiler` "virtual type" for our `virtual`.
- `name`: `numpy`, `ripgrep`, `openblas`, etc. package names for PURL. `c`, `cpp`, `rust`, etc. for our compiler virtual type

So then the logical follow-up would be: we need only well-defined virtual "types". I can only think of two: `compiler` and `interface`. The latter would be `openmp`, `mpi`, `blas`, `lapack`. I also had `virtual:ssl`, but that is actually wrong - while SSL is a protocol, it doesn't have a standard API/interface. `cryptography` supports multiple SSL libraries, but they're all forks of OpenSSL. So that example should simply use `pkg:generic/openssl`.

What can't really be expressed is "I need one of these three alternatives for function X". That doesn't fit in either `pkg` or `virtual` really, unless there's a syntax for expression `OR` or `fallback`.

**Comment by rgommers** [comment]
> Alternatively, how difficult would it be to convince the PURL authors to include a new `type` for these abstract requirements?

Yep, I had already opened gh-5 as an action for myself to do exactly this later today. That would be ideal I think.

**Comment by rgommers** [comment]
Done in https://github.com/package-url/purl-spec/issues/222

**Comment by pradyunsg** [comment]
> @pradyunsg WDYT?

I don't have enough understanding of the approaches we might have to take to resolve these virtual requirements to have a strong opinion formed on this. 😅 

My personal plan is to discuss the upcoming improvements that Spack is going to be making around compilers, and to survey the way Conda (and _maybe_ Fedora/Debian) are doing things to get a sense of what concretised variants of this would look like. Alternatively, we can expect those folks to engage during the PEP discussion (as they have) and help figuring out the answers to these questions.

**Comment by pradyunsg** [comment]
Honestly, as long as each "kind" of compiler maps to something unique, I'd be OK with whatever choice we make here. The current `compiler/{lang}` works well-enough IMO.

**Comment by rgommers** [comment]
Sounds like a good plan!

I think we'll also get a a better sense of prototyping some of the name mapping stuff. It's a matter of finding time for it, since it's not exactly trivial to do.

**Comment by rgommers** [comment]
> Given all that, I think we can close this issue as already handled. 

Thanks for the summary @jaimergp. Agreed with the conclusion.
